Ds3/Helpers/FileSystemHelpers.cs

```diff
@@ -0,0 +1,142 @@
+        /// be checked (can be a network drive)</param>
+        /// <param name="objectsToRead"></param>
+        /// <returns><see cref="DriveResult"/></returns>
+        public static DriveResult Check(string path, IEnumerable<Ds3Object> objectsToRead)
+        {
+            try
+            {
+                if (string.IsNullOrEmpty(path))
+                {
+                    var e = new ArgumentNullException(nameof(path));
+                    return new DriveResult(DriveStatus.Error, e.Message, e);
+                }
+
+                if (objectsToRead == null || !objectsToRead.Any())
+                {
+                    var e = new ArgumentNullException(nameof(objectsToRead));
+                    return new DriveResult(DriveStatus.Error, e.Message, e);
+                }
+
+                if (!Directory.Exists(path))
+                {
+                    return new DriveResult(DriveStatus.DirectoryNotExists, $"{path} does not exists.");
+                }
+
+                try
+                {
+                    if (!HavePermissions(path))
+                    {
+                        return new DriveResult(DriveStatus.NoPermission, $"User does not have permissions to access {path}.");
+                    }
+                }
+                catch (Exception e)
+                {
+                    return new DriveResult(DriveStatus.Error, e.Message, e);
+                }
+
+                try
+                {
+                    var freeBytes = DriveFreeBytes(path);
+                    var jobSize = (ulong)objectsToRead.Where(o => o.Size.HasValue).Sum(o => o.Size.Value);
+
+                    if (freeBytes < jobSize)
+                    {
+                        return new DriveResult(DriveStatus.NotEnoughSpace, $"Missing {jobSize - freeBytes} bytes.");
+                    }
+                }
+                catch (Exception e)
+                {
+                    return new DriveResult(DriveStatus.Error, e.Message, e);
+                }
+            }
+            catch (Exception e)
+            {
+                return new DriveResult(DriveStatus.Unknown, e.Message, e);
+            }
+
+            return new DriveResult(DriveStatus.Ok);
+        }
+
+        private static bool HavePermissions(string path)
+        {
+            try
+            {
+                using (File.Create(Path.Combine(path, Path.GetRandomFileName()), 1, FileOptions.DeleteOnClose))
+                {
+                    //do nothing
+                }
+                return true;
+            }
+            catch (UnauthorizedAccessException)
+            {
+                return false;
+            }
+        }
+
+        private static ulong DriveFreeBytes(string path)
+        {
+            if (!path.EndsWith("\\"))
+            {
+                path += '\\';
+            }
+
+            var driveInfo  = new DriveInfo(Path.GetPathRoot(path));
+
+            return (ulong)driveInfo.AvailableFreeSpace;
```

Comment:
ok, it looks like the other way is not better in my opinion
[http://stackoverflow.com/questions/9293227/how-to-check-if-ioexception-is-not-enough-disk-space-exception-type](url)
[https://msdn.microsoft.com/en-us/library/system.exception.hresult(v=vs.110).aspx](url)
@rpmoore let me know what do you think?